spec/lang/step.md

```diff
@@ -162,14 +161,50 @@ The functions `eval_un_op` and `eval_bin_op` are defined in [a separate file](op
 
 ```rust
 impl<M: Memory> Machine<M> {
-    fn eval_value(&mut self, ValueExpr::UnOp { operator, operand }: ValueExpr) -> NdResult<Value<M>> {
-        let operand = self.eval_value(operand)?;
-        ret(self.eval_un_op(operator, operand)?)
+    fn eval_value(&mut self, ValueExpr::UnOp { operator, operand }: ValueExpr) -> NdResult<(Value<M>, Type)> {
+        use lang::UnOp::*;
+
+        let (operand, _op_ty) = self.eval_value(operand)?;
+
+        let val = self.eval_un_op(operator, operand)?;
+        let ty = match operator {
```

Comment:
This function should not match on `operator`. Instead `eval_un_op` should work the type as well.

Same for binops.